golem-worker-service-base/src/gateway_execution/file_server_binding_handler.rs

```diff
@@ -173,6 +174,49 @@ impl DefaultFileServerBindingHandler {
             worker_service,
         }
     }
+
+    async fn get_component_metadata(
+        &self,
+        worker_detail: &WorkerDetails,
+    ) -> Result<Component, FileServerBindingError> {
+        // Two cases, we either have an existing worker or not (either not configured or not existing).
+        // If there is no worker we need use the lastest component version, if there is none we need to use the exact component version
+        // the worker is using. Not doing that would make the blob_storage optimization for read-only files visible to users.
+
+        let component_version = if let Some(worker_id) = worker_detail.worker_id() {
+            let worker_metadata = self
+                .worker_service
+                .get_metadata(&worker_id, empty_worker_metadata())
```

Comment:
So what happens if the worker has gone down by this time, and started using the latest component version?

Reply:
I know this approach is taken in other places. Just came to my mind, but sorry that I merged before getting an answer